transip/client.go

```diff
@@ -0,0 +1,118 @@
+package transip
+
+import (
+	"context"
+	"time"
+
+	"github.com/transip/gotransip"
+	transipdomain "github.com/transip/gotransip/domain"
+	"github.com/libdns/libdns"
+)
+
+func (p *Provider) setupRepository() error {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	if p.repository == nil {
+		client, err := gotransip.NewClient(gotransip.ClientConfiguration{
+			AccountName:	p.AccountName,
+			PrivateKeyPath:	p.PrivateKeyPath,
+		})
+		if err != nil {
+			return err
+		}
+		p.repository = transipdomain.Repository{Client: client}
+	}
+
+	return nil
+}
+
+func (p *Provider) getDNSEntries(ctx context.Context, domain string) ([]libdns.Record, error) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	p.setupRepository()
```

Comment:
Deadlock? setupRepository will also try to obtain the same lock, which won't be released until this function returns.

Reply:
Ah yes, that's stupid. I'm no mutex expert, would it be best to just have a mutex in each method and not in setupRepository?